src/rust/engine/graph/src/node.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::path::Path;
+
+use boxfuture::BoxFuture;
+use hashing::Digest;
+
+use petgraph::stable_graph;
+
+use Graph;
+
+// 2^32 Nodes ought to be more than enough for anyone!
+pub type EntryId = stable_graph::NodeIndex<u32>;
+
+///
+/// Defines executing a cacheable/memoizable step within the given NodeContext.
+///
+pub trait Node: Clone + Eq + Hash + Send + 'static {
+  type Context: NodeContext<Node = Self>;
+
+  type Item: Clone + Debug + Send + 'static;
+  type Error: NodeError;
+
+  fn run(self, context: Self::Context) -> BoxFuture<Self::Item, Self::Error>;
+
+  // TODO: Use a `Display` bound instead.
+  fn format(&self) -> String;
+
+  ///
+  /// If this Node represents an FS operation, returns its input Path.
```

Comment:
What kind of path? Absolute? Relative to build root?

Reply:
You trolled me into making this more generic.